### PR TITLE
lint: Speed up and fix flake8 checks

### PR DESCRIPTION
**Pull request by maflcko**
The checks have many issues:

* Some checks that could in theory hide bugs are not applied -> Fix them and apply them going forward
* Some checks are redundant Python 2 checks, or of low value -> Remove them
* The checks are slow -> Speed them up from ~10 seconds to about ~20 milliseconds

### CODE REVIEW

File: test/lint/test_runner/src/main.rs [hunk 1]
```diff
@@ -187,18 +187,76 @@ fn lint_subtree() -> LintResult {
...
+    .collect::<Vec<_>>();
     let files = check_output(
         git()
             .args(["ls-files", "--", "*.py"])
             .args(get_pathspecs_exclude_subtrees()),
     )?;
 
     let mut cmd = Command::new(bin_name);
-    cmd.arg("check").args(checks).args(files.lines());
+    cmd.args(["check", "--preview"])
```
Comment (kevkevinpal) [inline test/lint/test_runner/src/main.rs:257]: I am not sure how I feel about this `--preview` flag

According to their github docs in the [configuration section](https://github.com/astral-sh/ruff?tab=readme-ov-file#configuration)


> To opt in to the latest lint rules, formatter style changes, interface updates, and more, enable [preview mode](https://docs.astral.sh/ruff/rules/) by setting preview = true in your configuration file or passing --preview on the command line. Preview mode enables a collection of **unstable** features that may change prior to stabilization

File: test/lint/lint-python.py [hunk 2]
```diff
@@ -78,28 +78,20 @@
     'F632,'  # use ==/!= to compare str, bytes, and int literals
     'F701,'  # a break statement outside of a while or for loop
     'F702,'  # a continue statement outside of a while or for loop
-    'F703,'  # a continue statement in a finally block in a loop
     'F704,'  # a yield or yield from statement outside of a function
-    'F705,'  # a return statement with arguments inside a generator
     'F706,'  # a return statement outside of a function/method
```
Comment (davidgumberg) [inline test/lint/lint-python.py]: In commit: `lint: Remove python lint rules that are SyntaxError`

Nit: I believe all of F701-F707 result in `SyntaxError` and should be removed

<details>

<summary> The following examples resulted in SyntaxError's for Python 2.7.18 </summary>

f701.py:
```
def func(x):
    if x > 5:
        break

for i in range(10):
    func(i)        # <--- call function
```

f702.py:
```
def func(x):
    if x > 5:
        continue

for i in range(10):
    func(i)
```

f704.py:
```
step = 1

for x in range(0, 10, step):
  for y in range(0, 10, step):
    yield (x,y)
```

f706.py:
```
step = 1

for x in range(0, 10, step):
  for y in range(0, 10, step):
    return (x,y)
```

f707.py:
```
try:
    x = 1 / 0
except ZeroDivisionError:
    print("Divided by zero")
except:
    print("Generic exception")
except ValueError:
    print("Value error")
```

</details>

File: test/lint/lint-python.py [hunk 3]
```diff
@@ -82,7 +62,6 @@
...
     'F707,'  # an except: block as not the last exception handler
     'F811,'  # redefinition of unused name from line N
-    'F812,'  # list comprehension redefines 'foo' from line N
```
Comment (davidgumberg) [inline test/lint/lint-python.py:87]: This check doesn't seem like a terrible idea, but not supported by ruff.

File: test/lint/lint-python.py [hunk 4]
```diff
@@ -30,28 +30,8 @@
...
-    'E271,'  # multiple spaces after keyword
-    'E272,'  # multiple spaces before keyword
-    'E273,'  # tab after keyword
-    'E274,'  # tab before keyword
-    'E275,'  # missing whitespace after keyword
-    'E304,'  # blank lines found after function decorator
-    'E306,'  # expected 1 blank line before a nested definition
     'E401,'  # multiple imports on one line
     'E402,'  # module level import not at top of file
-    'E502,'  # the backslash is redundant between brackets
```
Comment (davidgumberg) [inline test/lint/lint-python.py:54]: This seems reasonable to me, but not available without the `--preview` argument

File: test/lint/lint-python.py [hunk 5]
```diff
@@ -30,28 +30,8 @@
...
-    'E224,'  # tab after operator
-    'E242,'  # tab after ','
-    'E266,'  # too many leading '#' for block comment
-    'E271,'  # multiple spaces after keyword
-    'E272,'  # multiple spaces before keyword
-    'E273,'  # tab after keyword
-    'E274,'  # tab before keyword
-    'E275,'  # missing whitespace after keyword
-    'E304,'  # blank lines found after function decorator
-    'E306,'  # expected 1 blank line before a nested definition
```
Comment (davidgumberg) [inline test/lint/lint-python.py:51]: In commit: [lint: Remove python whitespace and shadowing lint rules](https://github.com/bitcoin/bitcoin/pull/30723/commits/fafb0f62175798c5d5c174213c1ff1ccf5cea6ca)

+1 on dropping checks that result in `IndentError`

some of the other checks removed here seem reasonable to me e.g. E125, and E275, but not available/hidden behind preview flag in ruff so I'm ~0 on whether or not to remove the rest of them

File: test/lint/lint-python.py [hunk 6]
```diff
@@ -5,74 +5,25 @@
...
-    'E701,'  # multiple statements on one line (colon)
-    'E702,'  # multiple statements on one line (semicolon)
-    'E703,'  # statement ends with a semicolon
-    'E711,'  # comparison to None should be 'if cond is None:'
-    'E714,'  # test for object identity should be "is not"
-    'E721,'  # do not compare types, use "isinstance()"
-    'E722,'  # do not use bare 'except'
-    'E742,'  # do not define classes named "l", "O", or "I"
-    'E743,'  # do not define functions named "l", "O", or "I"
-    'E901,'  # SyntaxError: invalid syntax
```
Comment (davidgumberg) [inline test/lint/lint-python.py:64]: nit: this should've been removed in the [earlier commit](https://github.com/bitcoin/bitcoin/pull/30723/commits/66668f5292e0988461e5e34654c2b00dbbcb798e) that dropped syntaxerror checks

File: test/lint/test_runner/src/main.rs [hunk 7]
```diff
@@ -187,10 +187,54 @@ fn lint_subtree() -> LintResult {
...
+        "F402", // import module from line N shadowed by loop variable
+        "F403", // 'from foo_module import *' used; unable to detect undefined names
+        "F404", // future import(s) name after other statements
+        "F405", // foo_function may be undefined, or defined from star imports: bar_module
+        "F406", // "from module import *" only allowed at module level
+        "F407", // an undefined __future__ feature name was imported
+        "F601", // dictionary key name repeated with different values
+        "F602", // dictionary key variable name repeated with different values
+        "F621", // too many expressions in an assignment with star-unpacking
+        "F622", // two or more starred expressions in an assignment (a, *b, *c = d)
```
Comment (davidgumberg) [inline test/lint/test_runner/src/main.rs:216]: nit:
```python
# F622: two or more starred expressions in an assignment (a, *b, *c = d)
a, *b, *c = [1, 2, 3, 4]
```
results in a syntaxerror in Python 2.7.18

File: test/lint/test_runner/src/main.rs [hunk 8]
```diff
@@ -187,10 +187,54 @@ fn lint_subtree() -> LintResult {
...
+        "E701", // multiple statements on one line (colon)
+        "E702", // multiple statements on one line (semicolon)
+        "E703", // statement ends with a semicolon
+        "E711", // comparison to None should be 'if cond is None:'
+        "E714", // test for object identity should be "is not"
+        "E721", // do not compare types, use "isinstance()"
+        "E722", // do not use bare 'except'
+        "E742", // do not define classes named "l", "O", or "I"
+        "E743", // do not define functions named "l", "O", or "I"
+        "E902", // TokenError: EOF in multi-line string
```
Comment (davidgumberg) [inline test/lint/test_runner/src/main.rs:205]: nit:
```python
# E902: TokenError: EOF in multi-line string
"""
Unterminated string literal

```
results in a syntaxerror in Python 2.7.18

File: test/lint/test_runner/src/main.rs [hunk 9]
```diff
@@ -187,10 +187,54 @@ fn lint_subtree() -> LintResult {
...
+        "F841", // local variable 'foo' is assigned to but never used
+        "W191", // indentation contains tabs
+        "W291", // trailing whitespace
+        "W292", // no newline at end of file
+        "W293", // blank line contains whitespace
+        "W605", // invalid escape sequence "x"
+    ]
+    .iter()
+    .map(|c| format!("--select={}", c))
+    .collect::<Vec<_>>();
```
Comment (davidgumberg) [inline test/lint/test_runner/src/main.rs:237]: nit: ruff accepts the checks as comma separated values to --select=
```suggestion
    let checks = format!(
        "--select={}",
        [
            "B006", // mutable-argument-default
            "B008", // function-call-in-default-argument
            "E101", // indentation contains mixed spaces and tabs
            "E401", // multiple imports on one line
            "E402", // module level import not at top of file
            "E701", // multiple statements on one line (colon)
            "E702", // multiple statements on one line (semicolon)
            "E703", // statement ends with a semicolon
            "E711", // comparison to None should be 'if cond is None:'
            "E714", // test for object identity should be "is not"
            "E721", // do not compare types, use "isinstance()"
            "E722", // do not use bare 'except'
            "E742", // do not define classes named "l", "O", or "I"
            "E743", // do not define functions named "l", "O", or "I"
            "E902", // TokenError: EOF in multi-line string
            "F401", // module imported but unused
            "F402", // import module from line N shadowed by loop variable
            "F403", // 'from foo_module import *' used; unable to detect undefined names
            "F404", // future import(s) name after other statements
            "F405", // foo_function may be undefined, or defined from star imports: bar_module
            "F406", // "from module import *" only allowed at module level
            "F407", // an undefined __future__ feature name was imported
            "F601", // dictionary key name repeated with different values
            "F602", // dictionary key variable name repeated with different values
            "F621", // too many expressions in an assignment with star-unpacking
            "F622", // two or more starred expressions in an assignment (a, *b, *c = d)
            "F631", // assertion test is a tuple, which are always True
            "F632", // use ==/!= to compare str, bytes, and int literals
            "F701", // a break statement outside of a while or for loop
            "F702", // a continue statement outside of a while or for loop
            "F704", // a yield or yield from statement outside of a function
            "F706", // a return statement outside of a function/method
            "F707", // an except: block as not the last exception handler
            "F811", // redefinition of unused name from line N
            "F821", // undefined name 'Foo'
            "F822", // undefined name name in __all__
            "F823", // local variable name … referenced before assignment
            "F841", // local variable 'foo' is assigned to but never used
            "W191", // indentation contains tabs
            "W291", // trailing whitespace
            "W292", // no newline at end of file
            "W293", // blank line contains whitespace
            "W605", // invalid escape sequence "x"
        ]
        .join(",")
    );
```

**Comment by davidgumberg** [comment]
Tested and code reviewed ACK https://github.com/bitcoin/bitcoin/pull/30723/commits/fad05a7ae834af043d0cf47aeb6ea8faf2dda979

Looks good, left a few non-blocking nits about other checks that could be dropped since any reasonable version of python (>=  2.7.18) errors when these rules are violated, and stylistic nit about the format of arguments.

Tested with the following python file[^1] containing some of the disallowed linter violations:
https://gist.github.com/davidgumberg/1e2e43c7faeb5f96480142213adfaa3c

And our linter complained. In the end I was unable to trigger F821 and F822, as the linter always complains of F405 instead

[^1]: Generated initially with an LLM, and then manually inspected/corrected. It also appears that github gist swallows tabs and extraneous whitespace.

**Comment by davidgumberg** [comment]
review and tested reACK https://github.com/bitcoin/bitcoin/commit/fafdb7df34507eee735893aa871da6ae529e6372

The rebase drops checks for patterns that would result in `SyntaxError` in modern versions of python and makes a small improvement to the way the lint runner constructs the argument string for `ruff`.

**Comment by kevkevinpal** [comment]
ACK [fafdb7d](https://github.com/bitcoin/bitcoin/pull/30723/commits/fafdb7df34507eee735893aa871da6ae529e6372)

Removal of the unnecessary lint checks and the speed up of the checks are great updates!

**Comment by achow101** [comment]
ACK fafdb7df34507eee735893aa871da6ae529e6372

**Pull request merged**